### PR TITLE
Core/Phases: Fix SetInPhase not honoring zone phase

### PR DESCRIPTION
**Pull request by roc13x**
**Changes proposed:**
- Fix SetInPhase not honoring zone phases, only area
- Update area and zone phases on quest update

**Target branch(es):**
- [x] master

**Tests performed:**
- Compiled
- Tested in-game

### CODE REVIEW

File: src/server/game/Entities/Object/Object.cpp [hunk 1]
```diff
@@ -3028,11 +3028,13 @@ bool WorldObject::SetInPhase(uint32 id, bool update, bool apply)
         {
             // if area phase passes the condition we should not remove it (ie: if remove called from aura remove)
             // this however breaks the .mod phase command, you wont be able to remove any area based phases with it
-            if (std::vector<PhaseInfoStruct> const* phases = sObjectMgr->GetPhasesForArea(GetAreaId()))
-                for (PhaseInfoStruct const& phase : *phases)
-                    if (id == phase.Id)
-                        if (sConditionMgr->IsObjectMeetToConditions(this, phase.Conditions))
-                            return false;
+            PhaseInfo const& phases = sObjectMgr->GetAreaAndZonePhases();
+            for (PhaseInfo::const_iterator itr = phases.begin(); itr != phases.end(); ++itr)
```
Comment (Shauren) [inline src/server/game/Entities/Object/Object.cpp]: I don't like this - couldn't you make another store in ObjectMgr to lookup phase by id?

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -16005,7 +16005,7 @@ void Player::SendQuestUpdate(uint32 questId)
     }
 
     UpdateForQuestWorldObjects();
-    SendUpdatePhasing();
+    UpdateAreaAndZonePhase();
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: This was previously a target of discussion in another PR - by forcing the update here you break quests that should not update phase here

File: src/server/game/Entities/Object/Object.cpp (hunk 1 above)
Comment (roc13x) [inline src/server/game/Entities/Object/Object.cpp]: Wouldn't that require quite a bit of nearly-duplicate code to construct a similar store? Especially in ConditionMgr::addToPhases

Comment (Shauren) [inline src/server/game/Entities/Object/Object.cpp]: No, I mean something like `map<phaseId, vector<PhaseInfoStruct const*>>` in ObjectMgr (constructed in same place as `_phases`)

**Comment by Aokromes** [comment]
@Shauren this can be merged?

**Comment by Lopfest** [comment]
this PR fixes #21196 and almost #18190
please can you verify and add following code in `Object.cpp` in `SetInPhase` so it fixes also #18190?
https://pastebin.com/mASa9Vvs
you can also fix log message here
`TC_LOG_ERROR("sql.sql", "%s Area %u does not have phase %u.", cond->ToString().c_str(), cond->SourceGroup, cond->SourceEntry);`
cond->SourceGroup and cond->SourceEntry are swapped

**Comment by msoky** [comment]
any news?

**Comment by tkrokli** [comment]
Good question, but I don't see any recent development here.

BTW, is this PR affected by **Core/Object: fix logic in WorldObject::UpdateAreaAndZonePhase() #21214** and its commit https://github.com/TrinityCore/TrinityCore/commit/5b90538919cdf8b1984049bd203104128bec5bf2 ?

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Entities/Object/Object.cpp

**Comment by tkrokli** [comment]
Hardly any activity from @roc13x since October 2017 (view the user page).

**Comment by Shauren** [comment]
I started working on a bunch of phase related optimizations that will make this PR obsolete (you can see it here, accidentally pushed new files for this https://github.com/TrinityCore/TrinityCore/commit/bb718b557862057524778414ee728bf1923a14a4#diff-26a3ae191cac89dc294ec6b735af10ba)

**Comment by roc13x** [comment]
Ok good to know. Saves me reviewing this tonight since 5b90538919cdf8b1984049bd203104128bec5bf2 happened

**Pull request closed without merging**